roachtest: observe ctx cancellation in Run

tbg opened the pull request:
Before this patch, roachprod invocations would not observe
ctx cancellation. Or rather they would, but due to the usual
obscure passing of Stdout into the child process of roachprod
the Run call would not return until the child had finished.
As a result, the test would continue running, which is annoying
and also costs money.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/29178)
<!-- Reviewable:end -->

tschottdorf commented:
@petermattis this is just the first commit from #29174 which doesn't need to be blocked.

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though do you mind addressing the comment I left regarding `l.stdout == l.stderr`?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-:-LL-ryvpEAmt1E-zfirJ:b-emllff)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Oops, missed that - done (I think).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-:-LL-umKC-sN_5o0p3fOu:b-mfb0y5)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tschottdorf commented:
Let's hold off here so that I can try out the "panic escalation" idea (https://play.golang.com/p/0WUEJCAr0su)

tbg reviewed (commented):
@petermattis ready for another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-:-LL1-Awe95VU16Bv9Cpw:bof16i5)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Much prefer this approach. Minor question about the test you've added. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-:-LL11B2i9g_YgUsP6luy:b-x64euk)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/cluster.go, line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-LL1-go9Cdme6ecYdwUP:-LL1-go9Cdme6ecYdwUQ:b-udw0mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/01400e82c084fe14009df37966478a35b429134c/pkg/cmd/roachtest/cluster.go#L225)):*
> ```Go
> 
> 	debugStdoutBuffer, _ := circbuf.NewBuffer(1024)
> 	debugStderrBuffer, _ := circbuf.NewBuffer(1024)
> ```

Did you find this useful somewhere? Curious where it came up.
___
*[pkg/cmd/roachtest/cluster.go, line 1172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-LL101iA469yXTj_CFKa:-LL101iA469yXTj_CFKb:bo71loc) ([raw file](https://github.com/cockroachdb/cockroach/blob/01400e82c084fe14009df37966478a35b429134c/pkg/cmd/roachtest/cluster.go#L1172)):*
> ```Go
> 				// Note that the trick here is that we panicked explicitly below,
> 				// which somehow "overrides" the Goexit which is supposed to be
> 				// un-recoverable, but we do need to recover to return an error.
> ```

Oy! This is confusing.  Thanks for the comment.
___
*[pkg/cmd/roachtest/test_example_adversarial.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-LL11-_i0bS-0c-pZ2co:-LL11-_i0bS-0c-pZ2cp:bnnh9jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/01400e82c084fe14009df37966478a35b429134c/pkg/cmd/roachtest/test_example_adversarial.go#L34)):*
> ```Go
> }
> 
> func testHarness(ctx context.Context, t *test, c *cluster) {
> ```

Can this be an actual test (i.e `TestHarnest(t *testing.T)`)? Yeah, it forks `sleep` and `echo`, but that should be ok from a test.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/29178)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/cmd/roachtest/cluster.go, line 225 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-LL1-go9Cdme6ecYdwUP:-LL3wstg3MsLKreanW72:b-olsj76) ([raw file](https://github.com/cockroachdb/cockroach/blob/01400e82c084fe14009df37966478a35b429134c/pkg/cmd/roachtest/cluster.go#L225)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you find this useful somewhere? Curious where it came up.
</blockquote></details>

I think this will save more time than anything else in this PR - before you'd get something like "command XYZ returned exit status 1" and no details of what the process output. Take a look at the commit message and note how it captures the latest part of stdout/stderr. Basically I want to be able to look at the issue and see the failure instead of clicking through 12 teamcity pages and downloading up to five artifacts. I hope that this achieves that.
___
*[pkg/cmd/roachtest/cluster.go, line 1172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-LL101iA469yXTj_CFKa:-LL3xDwd-EE4PpOPElcJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01400e82c084fe14009df37966478a35b429134c/pkg/cmd/roachtest/cluster.go#L1172)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Oy! This is confusing.  Thanks for the comment.
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/test_example_adversarial.go, line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/29178#-LL11-_i0bS-0c-pZ2co:-LL3yPoh5TSyXXXgViv8:bypgphv) ([raw file](https://github.com/cockroachdb/cockroach/blob/01400e82c084fe14009df37966478a35b429134c/pkg/cmd/roachtest/test_example_adversarial.go#L34)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can this be an actual test (i.e `TestHarnest(t *testing.T)`)? Yeah, it forks `sleep` and `echo`, but that should be ok from a test.
</blockquote></details>

Done. Feels good to have decent coverage for this, but I seriously pity the next person who'll have to touch it. Probably gonna be me :-)


<!-- Sent from Reviewable.io -->

tbg committed:
roachtest: observe ctx cancellation in Run

Before this patch, roachprod invocations would not observe
ctx cancellation. Or rather they would, but due to the usual
obscure passing of Stdout into the child process of roachprod
the Run call would not return until the child had finished.
As a result, the test would continue running, which is annoying
and also costs money.

Also fixes up the handling of calling `c.t.Fatal` on a monitor
goroutine (using what is perhaps unspecified behavior of the
Go runtime). Anyway, the result is that you can do basically
whatever inside of a monitor and get away with it:

```go
m.Go(func(ctx context.Context) error {
	// Make sure the context cancellation works (not true prior to the PR
	// adding this test).
	return c.RunE(ctx, c.Node(1), "sleep", "2000")
})
m.Go(func(ctx context.Context) error {
	// This will call c.t.Fatal which also used to wreak havoc on the test
	// harness. Now it exits just fine (and all it took were some mean hacks).
	// Note how it will exit with stderr and stdout in the failure message,
	// which is extremely helpful.
	c.Run(ctx, c.Node(1), "echo foo && echo bar && notfound")
	return errors.New("impossible")
})
m.Wait()
```

now returns

```
--- FAIL: tpmc/w=1/nodes=3 (0.24s)
...,errgroup.go:58: /Users/tschottdorf/go/bin/roachprod run local:1 -- echo foo && echo bar && notfound returned:
	stderr:
	bash: notfound: command not found
	Error:  exit status 127

	stdout:
	foo
	bar
	: exit status 1
...,tpcc.go:661: Goexit() was called
FAIL
```

Release note: None

tschottdorf commented:
bors r=petermattis
